workflows4s-doobie/src/main/scala/workflows4s/doobie/sqlite/SqliteWorkflowStorage.scala

```diff
@@ -1,29 +1,38 @@
 package workflows4s.doobie.sqlite
 
+import cats.effect.IO
 import cats.effect.kernel.{Resource, Sync}
+import cats.effect.std.Semaphore
 import cats.syntax.all.*
 import doobie.*
 import doobie.implicits.*
+import doobie.util.transactor.Transactor
 import workflows4s.doobie.{ByteCodec, WorkflowStorage}
 import workflows4s.runtime.WorkflowInstanceId
 
-class SqliteWorkflowStorage[Event](eventCodec: ByteCodec[Event]) extends WorkflowStorage[Event] {
-  override def getEvents(id: WorkflowInstanceId): fs2.Stream[ConnectionIO, Event] =
+/** SQLite workflow storage.
+  *
+  * Uses a cats-effect Semaphore for locking since SQLite has restrictive concurrency (single writer). Each SQLite database file gets its own storage
+  * instance with its own lock.
+  */
+class SqliteWorkflowStorage[Event](xa: Transactor[IO], eventCodec: ByteCodec[Event], lock: Semaphore[IO]) extends WorkflowStorage[IO, Event] {
+  override def getEvents(id: WorkflowInstanceId): fs2.Stream[IO, Event] =
     sql"SELECT event_data FROM workflow_journal ORDER BY event_id"
       .query[Array[Byte]]
       .stream
       .evalMap(bytes => Sync[ConnectionIO].fromTry(eventCodec.read(IArray.unsafeFromArray(bytes))))
+      .transact(xa)
 
-  override def saveEvent(id: WorkflowInstanceId, event: Event): ConnectionIO[Unit] = {
+  override def saveEvent(id: WorkflowInstanceId, event: Event): IO[Unit] = {
     val bytes = IArray.genericWrapArray(eventCodec.write(event)).toArray
     sql"INSERT INTO workflow_journal (event_data) VALUES ($bytes)".update.run.void
+      .transact(xa)
   }
 
-  override def lockWorkflow(id: WorkflowInstanceId): Resource[ConnectionIO, Unit] = {
-    // SQLite locks the entire database with BEGIN IMMEDIATE
-    // This acquires a write lock on the database preventing other connections from writing
-    val acquire = sql"BEGIN IMMEDIATE".update.run.void
-    val release = sql"COMMIT".update.run.void
-    Resource.make(acquire)(_ => release)
-  }
+  override def lockWorkflow(id: WorkflowInstanceId): Resource[IO, Unit] = lock.permit
```

Comment:
This looks very wrong. Semaphoper is in-memory protectoection, while the old one was on the filesystem level, meaning two separate replicas couldnt access (and break) the database concurrently